[Python] 4/n Serialize OpenAI Model Parser

LastMileNewGrad opened the pull request:
[Python] 4/n Serialize OpenAI Model Parser



Modified serialize return type to list of Prompts.

Implemented serialize for OpenAI ModelParser.

Tested with api wrapper with the diff on top.

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/lastmile-ai/aiconfig/pull/33).
* __->__ #33
* #38
* #39
* #37

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +49,66 @@ def serialize(
-        # Cosntruct model settings, does not include prompt
-        prompt_model_metadata = {key: value for key, value in data.items() if key != "prompt"}
+        # Combine conversation data with any extra keyword args
+        conversation_data = {**data} | kwargs
```
Remove kwargs merging

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +49,66 @@ def serialize(
-            override_settings = {key: prompt_model_metadata[key] for key in override_keys}
+        if global_model_settings:
+            # Identify the settings that differ from global settings
+            override_settings = {
+                key: copy.deepcopy(conversation_data[key])
+                for key in conversation_data
+                if key not in global_model_settings
+            }
```
This isn't comparing the values of the keys, so if the global settings contain `temperature` but set to a 0.9, while input completion params contained `temperature` but set to 1, this will not consider that to be an override even though it should be.

saqadri replied:
Also check everywhere else in the code where setting overrides are being calculated, I am guessing the same problem exists elsewhere too.

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +49,66 @@ def serialize(
+        model_metadata.settings.pop("messages", None)
+
+        system_prompt = ""
+        user_prompt = ""
```
This is set to a string but then set to a Prompt object below, which is confusing. Maybe just set to None?

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +49,66 @@ def serialize(
+                    parameters=parameters,
+                )
+                serialized_prompts.append(user_prompt)
+                user_prompt = ""
```
```suggestion
                user_prompt = None
```

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +49,66 @@ def serialize(
+        system_prompt = ""
+        user_prompt = ""
+
+        for message in conversation_data["messages"]:
```
This doesn't handle `function` role

LastMileNewGrad commented:
Added multiple changes and refactors, to match ts implementation; old comments are stale or fixed

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +50,81 @@ def serialize(
-            parameters=parameters,
-            **kwargs
-        )
+                user_prompt = Prompt(
```
`prompt`

saqadri replied:
This hasn't been resolved. Rename variable to `prompt`

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +50,81 @@ def serialize(
+                        )
+                    ]
+                    if assistant_response
+                    else [],
```
`else None`

LastMileNewGrad replied:
addressed in comment below

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +50,81 @@ def serialize(
+                            output_type="execute_result",
+                            execution_count=None,
+                            data=assistant_response,
+                            metadata={},
```
`metadata=None`

LastMileNewGrad replied:
Metadata is defined to be of type `Dict`, pydantic will enforce this and not allow None type unless explicitly defined with `None` or `Optional`.

Will make a new diff on top of the stack to refactor to allow None as a default value. Will also update other types as per @jonathanlastmileai 's previous suggestion regarding setting defaults as `None` and handling from there.

saqadri commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +50,81 @@ def serialize(
+                )
+                serialized_prompts.append(user_prompt)
+            i += 1
+        return serialized_prompts
```
`prompts[prompts.length - 1].name = promptName;`

LastMileNewGrad commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +50,81 @@ def serialize(
-            # Compute the difference between the global model settings and the model settings from the input data
-            # If there is a difference, then we need to add the different model settings as overrides on the prompt's metadata
-            override_keys = set(prompt_model_metadata.keys()) - set(global_model_metadata.keys())
+        serialized_prompts = []
```
updated

LastMileNewGrad commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +50,81 @@ def serialize(
+        system_prompt = None
+        for message in conversation_data["messages"]:
+            if "role" in message and message["role"] == "system":
+                system_prompt = message["content"]
```
updated

LastMileNewGrad commented on python/src/aiconfig/default_parsers/openai.py:
```diff
@@ -74,40 +50,81 @@ def serialize(
+                        assistant_response = next_message
+                        i += 1
+                new_prompt_name = (
+                    f"{prompt_name}_{len(serialized_prompts) + len(ai_config.prompts) +1}"
```
changed to `len(prompt)` as it should be, not depending on the config

LastMileNewGrad commented:
> Add this to openai.ts in this diff as well.
> 
> ```
>  const input: PromptInput =
>           message.role === "user"
>             ? message.content ?? ""
>             : {
>                 ...message,
>               };
> ```

Updated ts 

- modified `serialize()` on the `Runtime`  object to require a prompt name, it is __not__ an optional attribute
- updated test/demo to reflect
